airbyte-cdk/java/airbyte-cdk/core/src/main/kotlin/io/airbyte/cdk/core/IntegrationCommand.kt

```diff
@@ -0,0 +1,165 @@
+    )
+    lateinit var configFile: Path
+
+    @CommandLine.Option(
+        names = ["--" + JavaBaseConstants.ARGS_CATALOG_KEY],
+        description =
+            [
+                JavaBaseConstants.ARGS_CATALOG_DESC,
+                "Required by the following commands: read, write",
+            ],
+    )
+    lateinit var catalogFile: Path
+
+    @CommandLine.Option(
+        names = ["--" + JavaBaseConstants.ARGS_STATE_KEY],
+        description =
+            [
+                JavaBaseConstants.ARGS_PATH_DESC,
+                "Required by the following commands: read",
+            ],
+    )
+    lateinit var stateFile: Optional<Path>
+
+    @CommandLine.Spec lateinit var commandSpec: CommandLine.Model.CommandSpec
+
+    override fun run() {
+        val result = execute(command)
+        result.onSuccess { airbyteMessage ->
+            airbyteMessage?.let { outputRecordCollector.accept(airbyteMessage) }
+        }
+        result.onFailure {
+            commandSpec.commandLine().usage(System.out)
+            logger.error(it) { "\nUnable to perform operation '$command'." }
+            // Many of the exceptions thrown are nested inside layers of RuntimeExceptions. An
+            // attempt is made
+            // to
+            // find the root exception that corresponds to a configuration error. If that does not
+            // exist, we
+            // just return the original exception.
+            ApmTraceUtils.addExceptionToTrace(it)
+            val rootThrowable = ConnectorExceptionUtil.getRootConfigError(Exception(it))
+            val displayMessage = ConnectorExceptionUtil.getDisplayMessage(rootThrowable)
+            // If the source connector throws a config error, a trace message with the relevant
+            // message should
+            // be surfaced.
+            if (ConnectorExceptionUtil.isConfigError(rootThrowable)) {
+                AirbyteTraceMessageUtility.emitConfigErrorTrace(it, displayMessage)
+            }
+            if (OperationType.CHECK.name.equals(command, true)) {
+                // Currently, special handling is required for the CHECK case since the user display
+                // information in
+                // the trace message is
+                // not properly surfaced to the FE. In the future, we can remove this and just throw
+                // an exception.
+                outputRecordCollector.accept(
+                    AirbyteMessage()
+                        .withType(AirbyteMessage.Type.CONNECTION_STATUS)
+                        .withConnectionStatus(
+                            AirbyteConnectionStatus()
+                                .withStatus(AirbyteConnectionStatus.Status.FAILED)
+                                .withMessage(displayMessage),
+                        ),
+                )
+            }
+        }
+
+        logger.info { "Completed integration: $connectorName" }
+    }
+
+    private fun execute(command: String): Result<AirbyteMessage?> {
```

Comment:
Why is the wrapped type `AirbyteMessage?`? Shouldn't it be `Sequence<AirbyteMessage>` instead? This is confusing and surprising.

Reply:
@postamar This whole thing needs another pass.  The existing code that has been used as inspiration doesn't distinguish between commands that return a single message and those that produce a sequence.  It feels like there should be two or three separate command interfaces:  one that returns void (e.g. write), one that returns a single message to be written to the output (e.g. check, spec, discover) and one that produces a sequence of messages (read).  I think that this is something that we can iterate on as we convert connectors and the pattern emerges more.

Reply:
This seems pretty fundamental. Let's introduce a union type with those subtypes right now:
1. exactly 0 messages
2. exactly 1 message
3. at least 1 message

Under the hood these subtypes can all be a `Sequence<AirbyteMessage>` or whatever; but we do also want to build sources with this code soon.

I don't see what value we gain from deferring this work but perhaps I'm just not seeing it.

Reply:
You've thought about this more than I have, evidently, I'll defer to your judgment. Let's defer this then.

Reply:
I'm going to reopen this because I'm not really satisfied with "let's see how this pans out". Why do an API before writing the connectors if we're not really thinking about what we're creating? To me, command should be a sealed type, and that should be parameterized with the return type

Reply:
If this was a simple change with enough experimentation that we knew what we wanted for the API, I would agree.  However, we don't have the data and the longer it takes for us to merge these changes, the longer it will take for us to get the API into the correct shape.  As it stands, this is a port of what exists now, which may not be ideal, but its what we have.  I am trying to limit the number of different approaches are part of this "version 0" of this work.  I'm all for fixing this once we have a better idea of how this may flesh out between sources and destinations.  Any change to this at this point is a premature optimization that may not survive contact with code in the connectors (specifically the sources that I am less familiar with) and is waste of time on top of a lot of wasted time up to this point.  At this point, we need to start to make progress on integrating this with a connector, even if that means we discover changes that we need.

Reply:
+1 to @jdpgrailsdev comment, lets not fix all the sins of the past here. If this is how it's working today lets stick with it and throw something in the backlog to address it later - there is value in delivering incremental change and testing this with real data.